src/vfs2.c

```diff
@@ -39,119 +40,241 @@
-  |                                                  ^                          ^                          |                         |
-  | sqlite3_open                                     | vfs2_apply_uncommitted   | vfs2_apply_uncommitted   | vfs2_{commit,unapply}   | vfs2_apply_uncommitted
-  v                                                  |                          |                          v                         |
-+----------------------------------------------------------------------------+  |                        +------------------------+  |
-|                                                                            |  +----------------------- |                        |  |
-|                                   EMPTY                                    |                           |         FLUSH          |  |
-|                                                                            | <------------------------ |                        |  |
-+----------------------------------------------------------------------------+   sqlite3_wal_checkpoint  +------------------------+  |
-  |                       ^                                                                                |                         |
-  | vfs2_commit_barrier   | sqlite3_wal_checkpoint                                                         |                         |
-  v                       |                                                                                |                         |
-+----------------------------------------------------------------------------+  vfs2_commit_barrier        |                         |
-|                                    BASE                                    | <---------------------------+                         |
-+----------------------------------------------------------------------------+                                                       |
-  |                       ^                          |                                                                               |
-  | sqlite3_step          | vfs2_unhide              +-------------------------------------------------------------------------------+
-  v                       |
-+----------------------+  |
-|        ACTIVE        |  |
-+----------------------+  |
-  |                       |
-  | COMMIT_PHASETWO       |
-  v                       |
-+----------------------+  |
-|        HIDDEN        |  |
-+----------------------+  |
-  |                       |
-  | vfs2_poll             |
-  v                       |
-+----------------------+  |
-|        POLLED        | -+
-+----------------------+
-
-*/
-
-static const struct sm_conf wtx_states[SM_STATES_MAX] = {
+/**
+ * Major transitions for this state machine:
+ *
+ *     +-----------CLOSED----------+
+ *     |                           |
+ * xOpen && no tx in WAL-cur   xOpen && tx in WAL-cur
```

Comment:
Does the "no tx" mean "no uncommitted tx"? There can be frames in the WAL from previous txs, right?

Reply:
It means "no transaction found by `walk_wal`", so transactions from previous WAL generations, invalid transactions, and those that are past the raft-provided stopping point are excluded. All such transactions in WAL-cur are considered uncommitted at startup until raft tells us otherwise by calling `vfs2_apply`.